src/interface/swarm_pack.hpp

```diff
@@ -0,0 +1,207 @@
+  using typename SwarmPackBase<TYPE>::contexts_t;
+  using typename SwarmPackBase<TYPE>::max_active_indices_t;
+
+  using SwarmPackBase<TYPE>::pack_;
+  using SwarmPackBase<TYPE>::bounds_;
+  using SwarmPackBase<TYPE>::contexts_;
+  using SwarmPackBase<TYPE>::max_active_indices_;
+  using SwarmPackBase<TYPE>::nvar_;
+  using SwarmPackBase<TYPE>::nblocks_;
+
+  explicit SwarmPack(const SwarmPackBase<TYPE> &spb) : SwarmPackBase<TYPE>(spb) {
+    if constexpr (sizeof...(Ts) != 0) {
+      static_assert(std::is_same<TYPE, typename GetDataType<Ts...>::value>::value,
+                    "Type mismatch in SwarmPack! When passing type-based variables as "
+                    "template argument to SwarmPack, ensure that the first template "
+                    "parameter is a data type (e.g., Real or int) that matches the "
+                    "data type of subsequent variable types!");
+    }
+  }
+
+  class Descriptor : public impl::SwarmPackDescriptor<TYPE> {
+   public:
+    Descriptor() = default;
+    explicit Descriptor(const impl::SwarmPackDescriptor<TYPE> &desc_in)
+        : impl::SwarmPackDescriptor<TYPE>(desc_in) {}
+
+    template <class T>
+    SwarmPack GetPack(T *pmd) const {
+      return SwarmPack(SwarmPackBase<TYPE>::GetPack(pmd, *this));
+    }
+
+    SparsePackIdxMap GetMap() const {
+      PARTHENON_REQUIRE(sizeof...(Ts) == 0,
+                        "Should not be getting an IdxMap for a type based pack");
+      return SwarmPackBase<TYPE>::GetIdxMap(*this);
+    }
+  };
+
+  const SwarmDeviceContext &GetContext(const int b = 0) const { return contexts_(b); }
+  const int &GetMaxActiveIndex(const int b = 0) const { return max_active_indices_(b); }
+
+  // Methods for getting parts of the shape of the pack
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetNBlocks() const { return nblocks_; }
+
+  // Bound overloads
+  KOKKOS_INLINE_FUNCTION int GetLowerBound(const int b) const { return bounds_(0, b, 0); }
+
+  KOKKOS_INLINE_FUNCTION int GetUpperBound(const int b) const {
+    return bounds_(1, b, nvar_ - 1);
+  }
+
+  KOKKOS_INLINE_FUNCTION int GetLowerBound(const int b, SwarmPackIdx idx) const {
+    static_assert(sizeof...(Ts) == 0, "Cannot create a string/type hybrid pack");
+    return bounds_(0, b, idx.VariableIdx());
+  }
+
+  KOKKOS_INLINE_FUNCTION int GetUpperBound(const int b, SwarmPackIdx idx) const {
+    static_assert(sizeof...(Ts) == 0, "Cannot create a string/type hybrid pack");
+    return bounds_(1, b, idx.VariableIdx());
+  }
+
+  template <class TIn, REQUIRES(IncludesType<TIn, Ts...>::value)>
+  KOKKOS_INLINE_FUNCTION int GetLowerBound(const int b, const TIn &) const {
+    const int vidx = GetTypeIdx<TIn, Ts...>::value;
+    return bounds_(0, b, vidx);
+  }
+
+  template <class TIn, REQUIRES(IncludesType<TIn, Ts...>::value)>
+  KOKKOS_INLINE_FUNCTION int GetUpperBound(const int b, const TIn &) const {
+    const int vidx = GetTypeIdx<TIn, Ts...>::value;
+    return bounds_(1, b, vidx);
+  }
```

Comment:
For swarms the bounds stuff probably doesn't need to depend on block, as swarms are not sparse.

Reply:
That said, it would be nice to add the `Contains` and the `Host` variants, so that one can check bounds on host and so that one can check whether or not a pack contains a particular swarm var.

Reply:
I did a lot of pruning for `SwarmPack`s.  It looks like some of the functionality I ripped out might be useful to others...

Reply:
I would remove the dependence on `b` but add the host variants and re-introduce `Contains`?

Reply:
Let's keep it for now, anticipating a merge b/w `SparsePack`s and `SwarmPack`s.  There was other significant cleanup to `bounds` in a recent commit following your advice.